File: solutions/python/spiral-matrix/1/spiral_matrix.py
def spiral_matrix(size):
    
    map = [[None for k in range(size)] for m in range(size)]
    direction = 0
    x, y = 0, 0

    def new_coordinates(y, x, direction):        
        def path_check(y, x):
            if direction == 0 and x + 1 < size: new_x, new_y = x + 1, y # right
            if direction == 1 and y + 1 < size: new_x, new_y = x, y + 1 # down
            if direction == 2 and x - 1 > -1: new_x, new_y = x - 1, y # left
            if direction == 3 and y - 1 > -1: new_x, new_y = x, y - 1 # up
            try:
                return new_y, new_x
            except:
                return y, x

        new_y, new_x = path_check(y, x)

        try:
            if map[new_y][new_x] == None:
                x, y = new_x, new_y
                return y, x, direction
        except:
            pass
        
        direction = (direction + 1) % 4
        y, x = path_check(y, x)

        return y, x, direction

    for k in range(size ** 2):
        map[y][x] = k + 1
        y, x, direction = new_coordinates(y, x, direction)

    return map    
    

File: solutions/cpp/spiral-matrix/1/spiral_matrix.cpp
#include "spiral_matrix.h"

namespace spiral_matrix {

// TODO: add your solution here
    std::vector<std::vector<unsigned int>> spiral_matrix(int size) {

        // create empty vector of vectors
        std::vector<std::vector<unsigned int>> output(size, std::vector<unsigned int>(size, 0));

        int x = 0;
        int y = 0;
        std::string orientation = "R";
        
        for (int i = 0; i < (size * size); i++) {
            if (orientation == "R") {
                if (x == size - 1) { // initial boundary
                    output[y][x] = i + 1;
                    y += 1;
                    orientation = "D";
                } else if (output[y][x+1] == 0) { // fill right
                    output[y][x] = i + 1;
                    x += 1;
                } else { // turn
                    output[y][x] = i + 1;
                    y += 1;
                    orientation = "D";
                }
            } else if (orientation == "D") {
                if (y == size - 1) { // initial boundary
                    output[y][x] = i + 1;
                    x -= 1;
                    orientation = "L";
                } else if (output[y+1][x] == 0) { // fill down
                    output[y][x] = i + 1;
                    y += 1;
                } else { // turn
                    output[y][x] = i + 1;
                    x -= 1;
                    orientation = "L";
                }
            } else if (orientation == "L") {
                if (x == 0) { // initial boundary
                    output[y][x] = i + 1;
                    y -= 1;
                    orientation = "U";
                } else if (output[y][x-1] == 0) { // fill left
                    output[y][x] = i + 1;
                    x -= 1;
                } else { // turn
                    output[y][x] = i + 1;
                    y -= 1;
                    orientation = "U";
                }
            } else if (orientation == "U") {
                if (y == 0) { // initial boundary
                    output[y][x] = i + 1;
                    x += 1;
                    orientation = "R";
                } else if (output[y-1][x] == 0) { // fill up
                    output[y][x] = i + 1;
                    y -= 1;
                } else { // turn
                    output[y][x] = i + 1;
                    x += 1;
                    orientation = "R";
                }
            }
        }
        
        return output;
    }
}  // namespace spiral_matrix
